File: X1/Alphabet Spam/as.c
// Authors: Felix Hedenström
//			Jonathan Rinnarv
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char * input[]){
	// Allocate a char array that allowes for 100 000 characters
	char * temp = malloc(100000 * sizeof(char));
	scanf("%s", temp); 	// Read the string
	int totalchars = 0;	// Counter of chars

	// Counters that keep track how often a character has appeared
	float whitespaceratio = 0;
	float lowercaseratio = 0;
	float uppercaseratio = 0;
	float symbolsratio = 0;

	// Go through the string until you find an end of line character
	while(*temp != 0 && *temp != 10 && *temp != 3){

		char temp_char = *temp; // Save a character from the string

		if(temp_char == 95)		// The character is a whitespace
			whitespaceratio++;
		else if(temp_char <= 122 && temp_char >= 97) // Lowercase character
			lowercaseratio++;
		else if(temp_char <= 90 && temp_char >= 65) // Uppercase character
			uppercaseratio++;
		else 										// Symbol
			symbolsratio++;
		
		totalchars++; 		// Add to the total
		temp = temp + 1; 	// Iterate the pointer
	
	}	
	// Print the ratios 
	printf("%.15f\n", whitespaceratio / totalchars);
	printf("%.15f\n", lowercaseratio / totalchars);
	printf("%.15f\n", uppercaseratio / totalchars);
	printf("%.15f\n", symbolsratio / totalchars);

}

File: X1/Alphabet Spam/as.hs
-- Authors: 	Felix Hedenström
-- 				Jonathan Rinnarv

module Main where
import Data.Char
import Data.Bool
import Data.List
import System.Environment

main = interact helpCounter

-- Initiates the count of all types of symbols
helpCounter :: String -> String
helpCounter s = helpShow(counter s 0 0 0 0 0) 


-- Creates a string from all elements in an array of doubles
helpShow :: [Double] -> String
helpShow [] = ""
helpShow (h:t) =  (show(h)++"\n")++helpShow t

-- Text - Whitespacecounter - lowercase - uppercase - symbol - total

-- Recursivly counts all whitespaces, lowercases, uppercases, symbols as well as the total number of symbols
counter :: [Char] -> Double -> Double -> Double -> Double -> Double -> [Double]

-- If there are no more characters, calculate the procentage for each symbol
counter [] w l u s tot = [w / tot, l / tot, u / tot , s  / tot]

-- If there are characters to read, go through them one by one recursivly
counter (h:t) w l u s tot 	
							| ord(h) == 10							= counter [] w l u s tot   			-- The character is an end of line character, end the counting
							| ord(h) == 95 							= counter t (w + 1) l u s (tot + 1) -- The character is a whitespace-character, add one to both the whitespacecounter and the totalcounter
							| (97 <= (ord(h))) && ((ord(h)) <= 122) = counter t w (l + 1) u s (tot + 1)	-- The character is a lowercase-character, add one to both the lowercasecounter and the totalcounter
							| (65 <= (ord(h))) && ((ord(h)) <= 90)	= counter t w l (u + 1) s (tot + 1) -- The character is a uppercase-character, add one to both the uppercasecounter and the totalcounter
							| otherwise 							= counter t w l u (s + 1) (tot + 1) -- The character is a symbol-character, add one to both the symbolcounter and the totalcounter
